File: python/node.py
#! /usr/bin/env python

class Node:
    depth_invocations = 0
    def __init__(self, children):
        self.children = children

    def depth(self):
        Node.depth_invocations += 1
        if not self.children:
            return 0
        max([c.depth() for c in self.children]) + 1
        #max_depth = 0
        #for c in self.children:
        #    max_depth = max(c.depth(), max_depth)
        #return max_depth + 1

a = Node([])
b = Node([a])
c = Node([b,a])
d = Node([a,b,c])
e = Node([a,b,c,d])

print "node depth: {0}, depth_invocations: {1}".format(e.depth(), Node.depth_invocations)

File: ruby/node.rb
#!/usr/bin/env ruby -w

class Node
  @@depth_invocations=0

  attr_reader :depth_invocations
  def initialize(children)
    @children = children
  end

  def depth
    @@depth_invocations += 1
    if @children.empty?
      return 0
    end
    @children.collect { |c| c.depth() }.max + 1
    # max_depth = 0
    # for node in @children
    # d = node.depth()
    # max_depth = [d, max_depth].max
    # end
    # return max_depth + 1
  end

  def self.depth_invocations
    @@depth_invocations
  end
end

a = Node.new([])
b = Node.new([a])
c = Node.new([a, b])
d = Node.new([a, b, c])

print "depth: ", d.depth, ", depth_invocations: ", Node.depth_invocations(), "\n"